src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIModelMapper.ChatCompletion.cs

```diff
@@ -0,0 +1,616 @@
+                out int value) ? value : null;
+
+            int? inputCachedTokenCount = additionalCounts.TryGetValue(
+                $"{nameof(ChatTokenUsage.InputTokenDetails)}.{nameof(ChatInputTokenUsageDetails.CachedTokenCount)}",
+                out value) ? value : null;
+
+            int? outputAudioTokenCount = additionalCounts.TryGetValue(
+                $"{nameof(ChatTokenUsage.OutputTokenDetails)}.{nameof(ChatOutputTokenUsageDetails.AudioTokenCount)}",
+                out value) ? value : null;
+
+            int? outputReasoningTokenCount = additionalCounts.TryGetValue(
+                $"{nameof(ChatTokenUsage.OutputTokenDetails)}.{nameof(ChatOutputTokenUsageDetails.ReasoningTokenCount)}",
+                out value) ? value : null;
+
+            if (inputAudioTokenCount is not null || inputCachedTokenCount is not null)
+            {
+                inputTokenUsageDetails = OpenAIChatModelFactory.ChatInputTokenUsageDetails(
+                    audioTokenCount: inputAudioTokenCount ?? 0,
+                    cachedTokenCount: inputCachedTokenCount ?? 0);
+            }
+
+            if (outputAudioTokenCount is not null || outputReasoningTokenCount is not null)
+            {
+                outputTokenUsageDetails = OpenAIChatModelFactory.ChatOutputTokenUsageDetails(
+                    audioTokenCount: outputAudioTokenCount ?? 0,
+                    reasoningTokenCount: outputReasoningTokenCount ?? 0);
+            }
+        }
+
+        return OpenAIChatModelFactory.ChatTokenUsage(
+            inputTokenCount: usageDetails.InputTokenCount ?? 0,
+            outputTokenCount: usageDetails.OutputTokenCount ?? 0,
+            totalTokenCount: usageDetails.TotalTokenCount ?? 0,
+            outputTokenDetails: outputTokenUsageDetails,
+            inputTokenDetails: inputTokenUsageDetails);
+    }
+
+    /// <summary>Converts an OpenAI role to an Extensions role.</summary>
+    private static ChatRole FromOpenAIChatRole(ChatMessageRole role) =>
+        role switch
+        {
+            ChatMessageRole.System => ChatRole.System,
+            ChatMessageRole.User => ChatRole.User,
+            ChatMessageRole.Assistant => ChatRole.Assistant,
+            ChatMessageRole.Tool => ChatRole.Tool,
+            _ => new ChatRole(role.ToString()),
+        };
+
+    /// <summary>Converts an Extensions role to an OpenAI role.</summary>
+    [return: NotNullIfNotNull("role")]
+    private static ChatMessageRole? ToOpenAIChatRole(ChatRole? role) =>
+        role switch
+        {
+            null => null,
+            _ when role == ChatRole.System => ChatMessageRole.System,
+            _ when role == ChatRole.User => ChatMessageRole.User,
+            _ when role == ChatRole.Assistant => ChatMessageRole.Assistant,
+            _ when role == ChatRole.Tool => ChatMessageRole.Tool,
+            _ => ChatMessageRole.System,
+        };
```

Comment:
Also, the use of switch here with when blocks seems like unnecessary ceremony. It could instead be:
```suggestion
        role is null ? null :
        role == ChatRole.System ? ChatMessageRole.System :
        role == ChatRole.User ? ChatMessageRole.User :
        role == ChatRole.Assistant ? ChatMessageRole.Assistant :
        role == ChatRole.Tool ? ChatMessageRole.Tool :
        ChatMessageRole.System;
```

Reply:
That was my original intent, but the analyzers in extensions find nested ternary operators to be disagreeable.

Reply:
Just suppress it at the top of the file. And subsequently I think we should disable some of these analyzers for the whole repo. @joperezr was amenable for any of these that are overreaching.